source/adapters/level_zero/kernel.hpp

```diff
@@ -15,16 +15,36 @@
 
 struct ur_kernel_handle_t_ : _ur_object {
   ur_kernel_handle_t_(bool OwnZeHandle, ur_program_handle_t Program)
-      : Context{nullptr}, Program{Program}, ZeKernel{nullptr},
-        SubmissionsCount{0}, MemAllocs{} {
+      : Context{nullptr}, Program{Program}, SubmissionsCount{0}, MemAllocs{} {
     OwnNativeHandle = OwnZeHandle;
   }
 
   ur_kernel_handle_t_(ze_kernel_handle_t Kernel, bool OwnZeHandle,
                       ur_context_handle_t Context)
-      : Context{Context}, Program{nullptr}, ZeKernel{Kernel},
-        SubmissionsCount{0}, MemAllocs{} {
+      : Context{Context}, Program{nullptr}, SubmissionsCount{0}, MemAllocs{} {
     OwnNativeHandle = OwnZeHandle;
+    createdFromNativeHandle = true;
+    ZeKernels.push_back({nullptr, Kernel});
+  }
+
+  ze_kernel_handle_t getZeKernel(ur_device_handle_t Device) {
+    assert(Device);
```

Comment:
The number of devices is basically set the moment we initialize the driver. In my event cache rewrite, I took advantage of that to give monotonically increasing ID's to the devices, allowing us to make this type of operation a direct lookup in an array.
https://github.com/pbalcer/unified-runtime/commit/147f0b4e4f2a63b2e6d0a19a7e690367f5c643bb#diff-89d38124dbcf1f91b31dc489a7e7c66bc0e6607ff77f2654810f74eae0826025R205